### PR TITLE
Revised specification notations.

### PR DESCRIPTION
**Pull request by gmalecha**
This is just some symbol pushing, nothing significant.

### CODE REVIEW

**Comment by gmalecha** [comment]
@jmgrosen Note, that this also fixes the proof in the `class` example, in case you are working on that.

File: theories/Auto/Notations.v [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  \with (x : nat)
+  \pre Exists y, [| x = 3 |] ** [| x = y |] -* empSP
+  \post Exists z : nat, empSP ** [| z = 0 |] -* empSP).
+Defined.
+*)
+
+Notation "'\pre' pre '\post' post" :=
+  ({| wpp_with := tdone
+    ; wpp_pre := pre
+    ; wpp_post := fun (r : val) => post |})
```
Comment (jmgrosen) [inline theories/Auto/Notations.v]: Will the `r` here affect anything? An underscore would be better here, but it's not worth worrying about if `r` is just ignored.

File: theories/Auto/Notations.v [hunk 2]
```diff
@@ -0,0 +1,57 @@
+Require Import Cpp.Sem.
+
+(* todo(gmm): these should be put in an interpretation scope *)
+Notation "'\with' x .. y '\pre' pre '\post' [ r ] post" :=
```
Comment (jmgrosen) [inline theories/Auto/Notations.v]: I'm not a fan of the notation to name the return value, but I can't think of anything better off the top of my head.

File: theories/Auto/Notations.v (hunk 1 above)
Comment (gmalecha) [reply theories/Auto/Notations.v]: Thanks for catching this, the `r` should definitely be an `_`!

File: theories/Auto/Notations.v (hunk 2 above)
Comment (gmalecha) [reply theories/Auto/Notations.v]: In papers we would usually write `r . P` but `.` is pretty special in Coq. I considered `\post (r) ...` which would be kind of like "a function call to the continuation" but I had a few issues with the parentheses notation. In both of these there is also the question about clashes with other notations. Originally, I considered

```coq
\pre [ ... ]
\post r [ ... ]
```

The nice thing about it being that the `[` `]` prevent clashes with notation within `pre` and `post`, I also tried `{` `}` but I couldn't get Coq to accept that.

File: tests/alloc/alloc_hpp_spec.v [hunk 3]
```diff
@@ -44,35 +44,30 @@ Section with_P.
     SConstructor "_Z5Alloc"
```
Comment (gmalecha) [inline tests/alloc/alloc_hpp_spec.v]: @jmgrosen the other thing I was wondering about is if there is a nice way to write the syntax for the function prototype, but we can do that in a different PR.

File: theories/Auto/Notations.v (hunk 2 above)
Comment (jmgrosen) [reply theories/Auto/Notations.v]: Why did you end up rejecting the bracket-surrounded notation? Did you just not want to add more noise?

In past projects, we've gotten around Coq being unhappy with `{` `}` by switching to `{{` `}}` :slightly_smiling_face:

File: theories/Auto/Notations.v [hunk 4]
```diff
@@ -0,0 +1,57 @@
+Require Import Cpp.Sem.
+
+(* todo(gmm): these should be put in an interpretation scope *)
```
Comment (gmalecha) [inline theories/Auto/Notations.v]: @jmgrosen I couldn't get the interpretation scopes to work, I'm not sure if you have any luck with this sort of thing, maybe I was just doing something foolish. We can address this in another PR.

File: theories/Auto/Notations.v [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+Notation "'\with' x .. y '\pre' pre '\post' [ r ] post" :=
+  ({| wpp_with := tcons (fun x => .. (tcons (fun y => tdone)) ..)
+    ; wpp_pre := fun x => .. (fun y => pre) ..
+    ; wpp_post := fun x => .. (fun y => (fun (r : val) => post)) .. |})
+  (at level 0, x closed binder, y closed binder, r ident, no associativity,
+   pre at level 200, post at level 200,
+   format "'[v'   '\with' '[' x .. y ']' '//' '\pre'  '[hv  '  pre  ']' '//' '\post'  [  r  ]  '[hv ' post  ']' ']'").
```
Comment (gmalecha) [inline theories/Auto/Notations.v]: For some reason, I also was not able to get these to print appropriately. I think that the notation might be too complex for Coq to pattern match against it.

Comment (jmgrosen) [reply theories/Auto/Notations.v]: I'll take a look at this after I've got my current work committed and can checkout this.

File: theories/Auto/Notations.v (hunk 2 above)
Comment (gmalecha) [reply theories/Auto/Notations.v]: Yeah. That might be a nice way to go. It seemed a little bit cluttered.

File: theories/Auto/Notations.v (hunk 5 above)
Comment (jmgrosen) [reply theories/Auto/Notations.v]: You got this working in the latest commit, right?

File: theories/Auto/Notations.v (hunk 2 above)
Comment (jmgrosen) [reply theories/Auto/Notations.v]: I just tried it out. I think I agree it seems too cluttered. It looks like FSCQ also has them bare.

Comment (jmgrosen) [reply theories/Auto/Notations.v]: I think I'm happy with the current notation for defining the return value, too, as long as we write it as `\post[r]` instead of `\post [ r ]` -- your current printer seems to reflect this, too.

File: theories/Auto/Notations.v [hunk 6]
```diff
@@ -10,15 +10,15 @@ Local Definition mkWpp {ts : tele} (f : teleF (mpred * (val -> mpred)) ts) : Wit
 Notation "'\with' x .. y '\pre' pre '\post' [ r ] post" :=
   (@mkWpp (tcons (fun x => .. (tcons (fun y => tdone)) ..))
          (fun x => .. (fun y => (pre, fun (r : val) => post)) ..))
-  (at level 0, x closed binder, y closed binder, r ident, no associativity,
+  (at level 10, x closed binder, y closed binder, r ident, no associativity,
```
Comment (jmgrosen) [inline theories/Auto/Notations.v]: I increased the level of these notations because at level 0, these aren't surrounded by parentheses anywhere, which looks pretty odd.

File: theories/Auto/Notations.v (hunk 4 above)
Comment (jmgrosen) [reply theories/Auto/Notations.v]: I investigated this a bit, and I think there are two issues:

1. `Bind Scope wpp_scope with WithPrePost.` doesn't work in arguments to `SFunction` etc. because Coq can't figure out (at parsing time, at least) that the result of the telescope has type `WithPrePost`.
2. `Arguments SFunction _ _ _%wpp.` doesn't work because ["The effect of the scope is limited to the argument itself. It does not propagate to subterms"](https://coq.inria.fr/refman/user-extensions/syntax-extensions.html#coq:cmd.arguments-scopes), and I think function bodies count as subterms.

I thought I figured out a hacky solution to (2) by re-defining function notation in `wpp_scope`, but that broke function notation in some other places. We could define our own function notation specifically for those places (`using_args this next data => \with ...`) but that seemed annoying.
